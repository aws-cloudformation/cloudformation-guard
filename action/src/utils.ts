import getConfig from './getConfig';

/**
 * Handle adding the root when a user supplies a path
 *
 * @function addRootPath
 * @param {string} path - A file path
 * @returns {string}
 */
export function addRootPath(path: string): string {
  const { path: root } = getConfig();
  if (path.startsWith('./')) {
    return `${root}${path.slice(1)}`;
  } else if (!path.startsWith(root)) {
    return `${root}/${path}`;
  }
  return path;
}

/**
 * Handle removing the root when a user supplies a path
 *
 * @function removeRootPath
 * @param {string} uri - File location URI
 * @returns {string}
 */
export function removeRootPath(uri: string): string {
  const { path } = getConfig();
  if (uri.startsWith(path)) {
    const pathWithoutRoot = uri.slice(path.length);
    return pathWithoutRoot.startsWith('/')
      ? pathWithoutRoot.slice(1)
      : pathWithoutRoot;
  }
  return uri;
}
