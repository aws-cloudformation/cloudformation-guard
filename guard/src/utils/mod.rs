#![allow(dead_code)]

pub mod reader;
pub mod writer;

#[derive(Debug, Clone)]
pub(crate) struct ReadCursor<'buffer> {
    line_num: usize,
    line_buffer: std::str::Lines<'buffer>,
    previous_lines: Vec<(usize, &'buffer str)>,
}

impl<'buffer> ReadCursor<'buffer> {
    pub(crate) fn new(buffer: &str) -> ReadCursor {
        ReadCursor {
            line_num: 0,
            line_buffer: buffer.lines(),
            previous_lines: Default::default(),
        }
    }

    pub(crate) fn next(&mut self) -> Option<(usize, &'buffer str)> {
        if self.line_num < self.previous_lines.len() {
            self.line_num += 1;
            return Some(self.previous_lines[self.line_num - 1]);
        }
        match self.line_buffer.next() {
            Some(line) => {
                self.line_num += 1;
                self.previous_lines.push((self.line_num, line));
                Some(self.previous_lines[self.line_num - 1])
            }
            None => None,
        }
    }

    #[cfg(test)]
    pub(crate) fn prev(&mut self) -> Option<(usize, &'buffer str)> {
        if self.line_num - 1 > 0 && !self.previous_lines.is_empty() {
            self.line_num -= 1;
            return Some(self.previous_lines[self.line_num]);
        }
        None
    }

    pub(crate) fn seek_line(&mut self, line: usize) -> Option<(usize, &'buffer str)> {
        if self.previous_lines.len() > line {
            self.line_num = line;
            return Some(self.previous_lines[self.line_num - 1]);
        }

        loop {
            match self.line_buffer.next() {
                Some(l) => {
                    self.line_num += 1;
                    self.previous_lines.push((self.line_num, l));
                    if self.line_num == line {
                        return Some(self.previous_lines[self.line_num - 1]);
                    }
                }
                None => return None,
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn read_cursor_test() {
        let resources = r#"
    Resources:
      s3:
        Type: AWS::S3::Bucket
      s3Policy:
        Type: AWS::S3::BucketPolicy
        Properties:
          PolicyDocument:
            Statement:
              Resource:
                Fn::Sub: "aws:arn:s3::${s3}""#;

        let mut cursor = ReadCursor::new(resources);
        while let Some(line) = cursor.next() {
            println!("{}.{}", line.0, line.1);
        }
        let prev = cursor.prev();
        assert!(prev.is_some());
        let prev = match prev {
            Some(p) => p,
            None => unreachable!(),
        };
        assert!(prev.1.contains("${s3}"));
        let _ = cursor.next();
        let mut lines = Vec::with_capacity(cursor.previous_lines.len());
        while let Some((line, prev)) = cursor.prev() {
            lines.insert(0, (line, prev));
            if prev.contains("s3Policy:") {
                break;
            }
        }
        for each in lines {
            println!("{}.{}", each.0, each.1);
        }
    }
}
