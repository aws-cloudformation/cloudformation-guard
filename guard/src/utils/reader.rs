use std::fs::File;
use std::io::{Cursor, Read, Stdin};

pub struct Reader {
    inner: ReadBuffer,
}

impl Default for Reader {
    fn default() -> Self {
        Self {
            inner: ReadBuffer::Stdin(std::io::stdin()),
        }
    }
}

impl Read for Reader {
    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
        match &mut self.inner {
            ReadBuffer::Stdin(stdin) => stdin.read(buf),
            ReadBuffer::Cursor(cursor) => cursor.read(buf),
            ReadBuffer::File(file) => file.read(buf),
        }
    }
}

impl Reader {
    pub fn new(stdin: ReadBuffer) -> Self {
        Self { inner: stdin }
    }
}

pub enum ReadBuffer {
    Stdin(Stdin),
    Cursor(Cursor<Vec<u8>>),
    File(File),
}
