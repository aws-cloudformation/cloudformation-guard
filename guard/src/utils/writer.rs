use crate::rules::errors::InternalError::{
    FromUtf8Error, IncompatibleWriterError, UnsupportedBufferError, UnsupportedOperationError,
};
use crate::Error;
use std::fs::File;
use std::io::{Read, Stderr, Stdout, Write};

#[derive(Debug)]
pub struct Writer {
    buffer: WriteBuffer,
    err: WriteBuffer,
}

impl Default for Writer {
    fn default() -> Self {
        Self {
            buffer: WriteBuffer::Stdout(std::io::stdout()),
            err: WriteBuffer::Stderr(std::io::stderr()),
        }
    }
}

impl Writer {
    pub fn new(buffer: WriteBuffer) -> crate::rules::Result<Self> {
        if buffer.is_err() {
            return Err(Error::from(IncompatibleWriterError(
                "Unable to use stderr as a regular buffer.".to_string(),
            )));
        }

        Ok(Self {
            buffer,
            err: WriteBuffer::Stderr(std::io::stderr()),
        })
    }

    pub fn new_with_err(buffer: WriteBuffer, err: WriteBuffer) -> crate::rules::Result<Self> {
        if buffer.is_err() {
            return Err(Error::from(IncompatibleWriterError(
                "Unable to use stderr as a regular buffer.".to_string(),
            )));
        }

        Ok(Self { buffer, err })
    }

    pub fn write_err(&mut self, s: String) -> std::io::Result<()> {
        writeln!(self.err, "{s}")
    }

    pub fn err_to_stripped(self) -> crate::rules::Result<String> {
        match self.err {
            WriteBuffer::Vec(vec) => String::from_utf8(strip_ansi_escapes::strip(vec)?)
                .map_err(|e| Error::from(FromUtf8Error(e))),
            WriteBuffer::File(mut file) => {
                let mut data = String::new();
                file.read_to_string(&mut data)
                    .expect("Unable to read from file");

                String::from_utf8(strip_ansi_escapes::strip(data)?)
                    .map_err(|e| Error::from(FromUtf8Error(e)))
            }
            WriteBuffer::Stdout(..) => Err(Error::from(UnsupportedOperationError(
                "Unable to call err_to_stripped() on a stdout buffer.".to_string(),
            ))),
            WriteBuffer::Stderr(..) => Err(Error::from(UnsupportedOperationError(
                "Unable to call err_to_stripped() on a stderr buffer.".to_string(),
            ))),
        }
    }

    pub fn into_string(self) -> crate::rules::Result<String> {
        self.buffer.into_string()
    }

    pub fn stripped(self) -> crate::rules::Result<String> {
        match self.buffer {
            WriteBuffer::Vec(vec) => {
                let stripped = strip_ansi_escapes::strip(vec)?;

                String::from_utf8(stripped).map_err(|e| Error::from(FromUtf8Error(e)))
            }
            WriteBuffer::File(mut file) => {
                let mut data = String::new();
                file.read_to_string(&mut data)
                    .expect("Unable to read from file");

                let stripped = strip_ansi_escapes::strip(data.into_bytes())?;

                String::from_utf8(stripped).map_err(|e| Error::from(FromUtf8Error(e)))
            }
            WriteBuffer::Stdout(..) => Err(Error::from(UnsupportedBufferError(
                "Unable to strip ANSI escapes from stdout buffer.".to_string(),
            ))),
            WriteBuffer::Stderr(..) => Err(Error::from(UnsupportedBufferError(
                "Unable to strip ANSI escapes from stderr buffer.".to_string(),
            ))),
        }
    }
}

impl Write for Writer {
    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
        self.buffer.write(String::from_utf8_lossy(buf).as_bytes())
    }

    fn flush(&mut self) -> std::io::Result<()> {
        self.buffer.flush()
    }
}

#[derive(Debug)]
pub enum WriteBuffer {
    Stdout(Stdout),
    Vec(Vec<u8>),
    File(File),
    Stderr(Stderr),
}

impl WriteBuffer {
    fn is_err(&self) -> bool {
        matches!(self, WriteBuffer::Stderr(_))
    }
    fn into_string(self) -> crate::rules::Result<String> {
        match self {
            WriteBuffer::Stdout(..) => Err(Error::from(UnsupportedOperationError(
                "Unable to call into_string() on a stdout buffer.".to_string(),
            ))),
            WriteBuffer::Stderr(..) => Err(Error::from(UnsupportedOperationError(
                "Unable to call into_string() on a stderr buffer.".to_string(),
            ))),
            WriteBuffer::Vec(vec) => {
                String::from_utf8(vec).map_err(|e| Error::from(FromUtf8Error(e)))
            }
            WriteBuffer::File(mut file) => {
                let mut data = String::new();
                file.read_to_string(&mut data)
                    .expect("Unable to read from file");

                Ok(data)
            }
        }
    }
}

impl Write for WriteBuffer {
    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
        match self {
            WriteBuffer::Stdout(stdout) => stdout.write(buf),
            WriteBuffer::Stderr(stderr) => stderr.write(buf),
            WriteBuffer::Vec(vec) => vec.write(buf),
            WriteBuffer::File(file) => file.write(buf),
        }
    }

    fn flush(&mut self) -> std::io::Result<()> {
        match self {
            WriteBuffer::Stdout(stdout) => stdout.flush(),
            WriteBuffer::Stderr(stderr) => stderr.flush(),
            WriteBuffer::Vec(vec) => vec.flush(),
            WriteBuffer::File(file) => file.flush(),
        }
    }
}
