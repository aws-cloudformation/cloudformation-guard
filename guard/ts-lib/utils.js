"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sanitizeSarifOutput = void 0;
const orderBy = require('lodash/orderBy');
function sanitizeProperty(obj, key, newValue) {
    if (typeof obj !== 'object' || obj === null) {
        return obj;
    }
    for (const prop in obj) {
        if (Object.prototype.hasOwnProperty.call(obj, prop)) {
            const value = obj[prop];
            if (prop === key && typeof value === 'string') {
                obj[prop] = newValue;
            }
            else if (typeof value === 'object' && value !== null) {
                obj[prop] = sanitizeProperty(value, key, newValue);
            }
        }
    }
    return obj;
}
const sanitizeSarifOutput = (result) => {
    const orderedArtifacts = orderBy(result.runs[0].artifacts, (item) => [item.location.uri, -item.location.uri]);
    result.runs[0].artifacts = orderedArtifacts;
    const orderedResults = orderBy(result.runs[0].results, (item) => [item.message.text, -item.message.text]);
    result.runs[0].results = orderedResults;
    const orderedLocations = orderBy(result.runs[0].results, (item) => [item.locations[0].physicalLocation.artifactLocation.uri, -item.locations[0]]);
    result.runs[0].results = orderedLocations;
    sanitizeProperty(result, 'uri', 'somePath');
    sanitizeProperty(result, 'version', 'x.x.x');
    sanitizeProperty(result, 'semanticVersion', 'x.x.x');
    sanitizeProperty(result, 'fullName', 'cfn-guard x.x.x');
    return result;
};
exports.sanitizeSarifOutput = sanitizeSarifOutput;
