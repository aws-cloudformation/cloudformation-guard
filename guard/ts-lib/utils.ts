import { SarifReport, SarifArtifact, SarifResult, SarifLocation, SarifDriver } from './index';
import { orderBy } from 'lodash';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function sanitizeProperty<T extends Record<string, any>>(obj: T, key: keyof T, newValue: string): T {
  if (typeof obj !== 'object' || obj === null) {
    return obj;
  }

  for (const prop in obj) {
    if (Object.prototype.hasOwnProperty.call(obj, prop)) {
      const value = obj[prop];

      if (prop === key && typeof value === 'string') {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        (obj as any)[prop] = newValue;
      } else if (typeof value === 'object' && value !== null) {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        (obj as any)[prop] = sanitizeProperty(value, key, newValue);
      }
    }
  }

  return obj;
}

const sanitizeSarifOutput = (result: SarifReport): SarifReport => {
  const orderedArtifacts = orderBy(result.runs[0].artifacts, (item: SarifArtifact) => [item.location.uri, -item.location.uri]);
  result.runs[0].artifacts = orderedArtifacts;

  const orderedResults = orderBy(result.runs[0].results, (item: SarifResult) => [item.message.text, -item.message.text]);
  result.runs[0].results = orderedResults;

  const orderedLocations = orderBy(result.runs[0].results, (item: SarifResult) => [item.locations[0].physicalLocation.artifactLocation.uri, -item.locations[0]]);
  result.runs[0].results = orderedLocations;

  sanitizeProperty(result as unknown as SarifLocation, 'uri', 'somePath');
  sanitizeProperty(result, 'version', 'x.x.x');
  sanitizeProperty(result as unknown as SarifDriver, 'semanticVersion', 'x.x.x');
  sanitizeProperty(result as unknown as SarifDriver, 'fullName', 'cfn-guard x.x.x');

  return result;
};

export { sanitizeSarifOutput };
