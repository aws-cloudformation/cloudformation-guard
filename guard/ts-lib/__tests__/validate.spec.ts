import { validate } from '../index.js'
import { sanitizeSarifOutput } from '../utils';
const path = require('path')

describe('validate', () => {
  it('should match the snapshot', async () => {
    const result = await validate({
      rulesPath: path.resolve(__dirname, '../../resources/validate/rules-dir'),
      dataPath: path.resolve(__dirname, '../../resources/validate/data-dir'),
    })

    expect(sanitizeSarifOutput(result)).toMatchSnapshot()
  })
})
