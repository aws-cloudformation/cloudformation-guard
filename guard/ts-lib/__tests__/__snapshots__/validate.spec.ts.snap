// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate should match the snapshot 1`] = `
{
  "$schema": "https://docs.oasis-open.org/sarif/sarif/v2.1.0/errata01/os/schemas/sarif-schema-2.1.0.json",
  "runs": [
    {
      "artifacts": [
        {
          "location": {
            "uri": "somePath",
          },
        },
        {
          "location": {
            "uri": "somePath",
          },
        },
        {
          "location": {
            "uri": "somePath",
          },
        },
        {
          "location": {
            "uri": "somePath",
          },
        },
        {
          "location": {
            "uri": "somePath",
          },
        },
      ],
      "results": [
        {
          "level": "error",
          "locations": [
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "somePath",
                },
                "region": {
                  "startColumn": 1,
                  "startLine": 1,
                },
              },
            },
          ],
          "message": {
            "text": "Check was not compliant as property value [Path=/NotAwsAccessKey[L:0,C:17] Value="AKIAIOSFODNN7EXAMPLE"] equal to value [Path=[L:0,C:0] Value="/(?<![A-Z0-9])[A-Z0-9]{20}(?![A-Z0-9])/"]. ",
          },
          "ruleId": "ADVANCED_REGEX_NEGATIVE_LOOKBEHIND_RULE",
        },
        {
          "level": "error",
          "locations": [
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "somePath",
                },
                "region": {
                  "startColumn": 1,
                  "startLine": 1,
                },
              },
            },
          ],
          "message": {
            "text": "Check was not compliant as property value [Path=/NotSecretAccessKey[L:1,C:20] Value="wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY"] equal to value [Path=[L:0,C:0] Value="/(?<![A-Za-z0-9\\/+=])[A-Za-z0-9\\/+=]{40}(?![A-Za-z0-9\\/+=])/"]. ",
          },
          "ruleId": "ADVANCED_REGEX_NEGATIVE_LOOKBEHIND_RULE",
        },
        {
          "level": "error",
          "locations": [
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "somePath",
                },
                "region": {
                  "startColumn": 6,
                  "startLine": 8,
                },
              },
            },
          ],
          "message": {
            "text": "Check was not compliant as property [LoggingConfiguration] is missing. Value traversed to [Path=/Resources/MyBucket/Properties[L:8,C:6] Value={"PublicAccessBlockConfiguration":{"BlockPublicAcls":true,"BlockPublicPolicy":true,"IgnorePublicAcls":true,"RestrictPublicBuckets":true},"BucketEncryption":{"ServerSideEncryptionConfiguration":[{"ServerSideEncryptionByDefault":{"SSEAlgorithm":"AES256"}}]},"VersioningConfiguration":{"Status":"Enabled"}}]. 
    Violation: S3 Bucket Logging needs to be configured to enable logging.
    Fix: Set the S3 Bucket property LoggingConfiguration to start logging into S3 bucket.
  ",
          },
          "ruleId": "S3_BUCKET_LOGGING_ENABLED",
        },
        {
          "level": "error",
          "locations": [
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "somePath",
                },
                "region": {
                  "startColumn": 1,
                  "startLine": 4,
                },
              },
            },
          ],
          "message": {
            "text": "Check was not compliant as property [NotAwsAccessKey] to compare from is missing. Value traversed to [Path=[L:4,C:0] Value={"Resources":{"MyBucket":{"Type":"AWS::S3::Bucket","Properties":{"PublicAccessBlockConfiguration":{"BlockPublicAcls":true,"BlockPublicPolicy":true,"IgnorePublicAcls":true,"RestrictPublicBuckets":true},"BucketEncryption":{"ServerSideEncryptionConfiguration":[{"ServerSideEncryptionByDefault":{"SSEAlgorithm":"AES256"}}]},"VersioningConfiguration":{"Status":"Enabled"}}}}}]. ",
          },
          "ruleId": "ADVANCED_REGEX_NEGATIVE_LOOKBEHIND_RULE",
        },
        {
          "level": "error",
          "locations": [
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "somePath",
                },
                "region": {
                  "startColumn": 1,
                  "startLine": 4,
                },
              },
            },
          ],
          "message": {
            "text": "Check was not compliant as property [NotSecretAccessKey] to compare from is missing. Value traversed to [Path=[L:4,C:0] Value={"Resources":{"MyBucket":{"Type":"AWS::S3::Bucket","Properties":{"PublicAccessBlockConfiguration":{"BlockPublicAcls":true,"BlockPublicPolicy":true,"IgnorePublicAcls":true,"RestrictPublicBuckets":true},"BucketEncryption":{"ServerSideEncryptionConfiguration":[{"ServerSideEncryptionByDefault":{"SSEAlgorithm":"AES256"}}]},"VersioningConfiguration":{"Status":"Enabled"}}}}}]. ",
          },
          "ruleId": "ADVANCED_REGEX_NEGATIVE_LOOKBEHIND_RULE",
        },
        {
          "level": "error",
          "locations": [
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "somePath",
                },
                "region": {
                  "startColumn": 6,
                  "startLine": 13,
                },
              },
            },
          ],
          "message": {
            "text": "Check was not compliant as property [LoggingConfiguration] is missing. Value traversed to [Path=/Resources/MyBucket/Properties[L:13,C:6] Value={"BucketEncryption":{"ServerSideEncryptionConfiguration":[{"ServerSideEncryptionByDefault":{"SSEAlgorithm":"AES256"}}]},"VersioningConfiguration":{"Status":"Enabled"}}]. 
    Violation: S3 Bucket Logging needs to be configured to enable logging.
    Fix: Set the S3 Bucket property LoggingConfiguration to start logging into S3 bucket.
  ",
          },
          "ruleId": "S3_BUCKET_LOGGING_ENABLED",
        },
        {
          "level": "error",
          "locations": [
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "somePath",
                },
                "region": {
                  "startColumn": 1,
                  "startLine": 4,
                },
              },
            },
          ],
          "message": {
            "text": "Check was not compliant as property [NotAwsAccessKey] to compare from is missing. Value traversed to [Path=[L:4,C:0] Value={"Resources":{"MyBucket":{"Type":"AWS::S3::Bucket","Properties":{"BucketEncryption":{"ServerSideEncryptionConfiguration":[{"ServerSideEncryptionByDefault":{"SSEAlgorithm":"AES256"}}]},"VersioningConfiguration":{"Status":"Enabled"}}}}}]. ",
          },
          "ruleId": "ADVANCED_REGEX_NEGATIVE_LOOKBEHIND_RULE",
        },
        {
          "level": "error",
          "locations": [
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "somePath",
                },
                "region": {
                  "startColumn": 1,
                  "startLine": 4,
                },
              },
            },
          ],
          "message": {
            "text": "Check was not compliant as property [NotSecretAccessKey] to compare from is missing. Value traversed to [Path=[L:4,C:0] Value={"Resources":{"MyBucket":{"Type":"AWS::S3::Bucket","Properties":{"BucketEncryption":{"ServerSideEncryptionConfiguration":[{"ServerSideEncryptionByDefault":{"SSEAlgorithm":"AES256"}}]},"VersioningConfiguration":{"Status":"Enabled"}}}}}]. ",
          },
          "ruleId": "ADVANCED_REGEX_NEGATIVE_LOOKBEHIND_RULE",
        },
        {
          "level": "error",
          "locations": [
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "somePath",
                },
                "region": {
                  "startColumn": 6,
                  "startLine": 13,
                },
              },
            },
          ],
          "message": {
            "text": "Check was not compliant as property [PublicAccessBlockConfiguration.BlockPublicAcls] to compare from is missing. Value traversed to [Path=/Resources/MyBucket/Properties[L:13,C:6] Value={"BucketEncryption":{"ServerSideEncryptionConfiguration":[{"ServerSideEncryptionByDefault":{"SSEAlgorithm":"AES256"}}]},"VersioningConfiguration":{"Status":"Enabled"}}]. ",
          },
          "ruleId": "S3_BUCKET_PUBLIC_READ_PROHIBITED",
        },
        {
          "level": "error",
          "locations": [
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "somePath",
                },
                "region": {
                  "startColumn": 6,
                  "startLine": 13,
                },
              },
            },
          ],
          "message": {
            "text": "Check was not compliant as property [PublicAccessBlockConfiguration.BlockPublicPolicy] to compare from is missing. Value traversed to [Path=/Resources/MyBucket/Properties[L:13,C:6] Value={"BucketEncryption":{"ServerSideEncryptionConfiguration":[{"ServerSideEncryptionByDefault":{"SSEAlgorithm":"AES256"}}]},"VersioningConfiguration":{"Status":"Enabled"}}]. ",
          },
          "ruleId": "S3_BUCKET_PUBLIC_READ_PROHIBITED",
        },
        {
          "level": "error",
          "locations": [
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "somePath",
                },
                "region": {
                  "startColumn": 6,
                  "startLine": 13,
                },
              },
            },
          ],
          "message": {
            "text": "Check was not compliant as property [PublicAccessBlockConfiguration.IgnorePublicAcls] to compare from is missing. Value traversed to [Path=/Resources/MyBucket/Properties[L:13,C:6] Value={"BucketEncryption":{"ServerSideEncryptionConfiguration":[{"ServerSideEncryptionByDefault":{"SSEAlgorithm":"AES256"}}]},"VersioningConfiguration":{"Status":"Enabled"}}]. ",
          },
          "ruleId": "S3_BUCKET_PUBLIC_READ_PROHIBITED",
        },
        {
          "level": "error",
          "locations": [
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "somePath",
                },
                "region": {
                  "startColumn": 6,
                  "startLine": 13,
                },
              },
            },
          ],
          "message": {
            "text": "Check was not compliant as property [PublicAccessBlockConfiguration.RestrictPublicBuckets] to compare from is missing. Value traversed to [Path=/Resources/MyBucket/Properties[L:13,C:6] Value={"BucketEncryption":{"ServerSideEncryptionConfiguration":[{"ServerSideEncryptionByDefault":{"SSEAlgorithm":"AES256"}}]},"VersioningConfiguration":{"Status":"Enabled"}}]. 
    Violation: S3 Bucket Public Write Access controls need to be restricted.
    Fix: Set S3 Bucket PublicAccessBlockConfiguration properties for BlockPublicAcls, BlockPublicPolicy, IgnorePublicAcls, RestrictPublicBuckets parameters to true.
  ",
          },
          "ruleId": "S3_BUCKET_PUBLIC_READ_PROHIBITED",
        },
        {
          "level": "error",
          "locations": [
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "somePath",
                },
                "region": {
                  "startColumn": 6,
                  "startLine": 13,
                },
              },
            },
          ],
          "message": {
            "text": "Check was not compliant as property [PublicAccessBlockConfiguration] is missing. Value traversed to [Path=/Resources/MyBucket/Properties[L:13,C:6] Value={"BucketEncryption":{"ServerSideEncryptionConfiguration":[{"ServerSideEncryptionByDefault":{"SSEAlgorithm":"AES256"}}]},"VersioningConfiguration":{"Status":"Enabled"}}]. ",
          },
          "ruleId": "S3_BUCKET_PUBLIC_READ_PROHIBITED",
        },
        {
          "level": "error",
          "locations": [
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "somePath",
                },
                "region": {
                  "startColumn": 6,
                  "startLine": 8,
                },
              },
            },
          ],
          "message": {
            "text": "Check was not compliant as property [LoggingConfiguration] is missing. Value traversed to [Path=/Resources/MyBucket/Properties[L:8,C:6] Value={"PublicAccessBlockConfiguration":{"BlockPublicAcls":true,"BlockPublicPolicy":true,"IgnorePublicAcls":true,"RestrictPublicBuckets":true},"BucketEncryption":{"ServerSideEncryptionConfiguration":[{"ServerSideEncryptionByDefault":{"SSEAlgorithm":"AES256"}}]},"VersioningConfiguration":{"Status":"Enabled"}}]. 
    Violation: S3 Bucket Logging needs to be configured to enable logging.
    Fix: Set the S3 Bucket property LoggingConfiguration to start logging into S3 bucket.
  ",
          },
          "ruleId": "S3_BUCKET_LOGGING_ENABLED",
        },
        {
          "level": "error",
          "locations": [
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "somePath",
                },
                "region": {
                  "startColumn": 1,
                  "startLine": 4,
                },
              },
            },
          ],
          "message": {
            "text": "Check was not compliant as property [NotAwsAccessKey] to compare from is missing. Value traversed to [Path=[L:4,C:0] Value={"Resources":{"MyBucket":{"Type":"AWS::S3::Bucket","Properties":{"PublicAccessBlockConfiguration":{"BlockPublicAcls":true,"BlockPublicPolicy":true,"IgnorePublicAcls":true,"RestrictPublicBuckets":true},"BucketEncryption":{"ServerSideEncryptionConfiguration":[{"ServerSideEncryptionByDefault":{"SSEAlgorithm":"AES256"}}]},"VersioningConfiguration":{"Status":"Enabled"}}}}}]. ",
          },
          "ruleId": "ADVANCED_REGEX_NEGATIVE_LOOKBEHIND_RULE",
        },
        {
          "level": "error",
          "locations": [
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "somePath",
                },
                "region": {
                  "startColumn": 1,
                  "startLine": 4,
                },
              },
            },
          ],
          "message": {
            "text": "Check was not compliant as property [NotSecretAccessKey] to compare from is missing. Value traversed to [Path=[L:4,C:0] Value={"Resources":{"MyBucket":{"Type":"AWS::S3::Bucket","Properties":{"PublicAccessBlockConfiguration":{"BlockPublicAcls":true,"BlockPublicPolicy":true,"IgnorePublicAcls":true,"RestrictPublicBuckets":true},"BucketEncryption":{"ServerSideEncryptionConfiguration":[{"ServerSideEncryptionByDefault":{"SSEAlgorithm":"AES256"}}]},"VersioningConfiguration":{"Status":"Enabled"}}}}}]. ",
          },
          "ruleId": "ADVANCED_REGEX_NEGATIVE_LOOKBEHIND_RULE",
        },
        {
          "level": "error",
          "locations": [
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "somePath",
                },
                "region": {
                  "startColumn": 6,
                  "startLine": 8,
                },
              },
            },
          ],
          "message": {
            "text": "Check was not compliant as property [LoggingConfiguration] is missing. Value traversed to [Path=/Resources/MyBucket/Properties[L:8,C:6] Value={"PublicAccessBlockConfiguration":{"BlockPublicAcls":true,"BlockPublicPolicy":true,"IgnorePublicAcls":true,"RestrictPublicBuckets":true},"BucketEncryption":"","VersioningConfiguration":{"Status":"Enabled"}}]. 
    Violation: S3 Bucket Logging needs to be configured to enable logging.
    Fix: Set the S3 Bucket property LoggingConfiguration to start logging into S3 bucket.
  ",
          },
          "ruleId": "S3_BUCKET_LOGGING_ENABLED",
        },
        {
          "level": "error",
          "locations": [
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "somePath",
                },
                "region": {
                  "startColumn": 1,
                  "startLine": 4,
                },
              },
            },
          ],
          "message": {
            "text": "Check was not compliant as property [NotAwsAccessKey] to compare from is missing. Value traversed to [Path=[L:4,C:0] Value={"Resources":{"MyBucket":{"Type":"AWS::S3::Bucket","Properties":{"PublicAccessBlockConfiguration":{"BlockPublicAcls":true,"BlockPublicPolicy":true,"IgnorePublicAcls":true,"RestrictPublicBuckets":true},"BucketEncryption":"","VersioningConfiguration":{"Status":"Enabled"}}}}}]. ",
          },
          "ruleId": "ADVANCED_REGEX_NEGATIVE_LOOKBEHIND_RULE",
        },
        {
          "level": "error",
          "locations": [
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "somePath",
                },
                "region": {
                  "startColumn": 1,
                  "startLine": 4,
                },
              },
            },
          ],
          "message": {
            "text": "Check was not compliant as property [NotSecretAccessKey] to compare from is missing. Value traversed to [Path=[L:4,C:0] Value={"Resources":{"MyBucket":{"Type":"AWS::S3::Bucket","Properties":{"PublicAccessBlockConfiguration":{"BlockPublicAcls":true,"BlockPublicPolicy":true,"IgnorePublicAcls":true,"RestrictPublicBuckets":true},"BucketEncryption":"","VersioningConfiguration":{"Status":"Enabled"}}}}}]. ",
          },
          "ruleId": "ADVANCED_REGEX_NEGATIVE_LOOKBEHIND_RULE",
        },
        {
          "level": "error",
          "locations": [
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "somePath",
                },
                "region": {
                  "startColumn": 23,
                  "startLine": 13,
                },
              },
            },
          ],
          "message": {
            "text": "Check was not compliant as property [ServerSideEncryptionConfiguration[*].ServerSideEncryptionByDefault.SSEAlgorithm] to compare from is missing. Value traversed to [Path=/Resources/MyBucket/Properties/BucketEncryption[L:13,C:23] Value=""]. 
    Violation: S3 Bucket must enable server-side encryption.
    Fix: Set the S3 Bucket property BucketEncryption.ServerSideEncryptionConfiguration.ServerSideEncryptionByDefault.SSEAlgorithm to either "aws:kms" or "AES256"
  ",
          },
          "ruleId": "S3_BUCKET_SERVER_SIDE_ENCRYPTION_ENABLED",
        },
      ],
      "tool": {
        "driver": {
          "downloadUri": "https://github.com/aws-cloudformation/cloudformation-guard",
          "fullName": "cfn-guard x.x.x",
          "informationUri": "https://github.com/aws-cloudformation/cloudformation-guard",
          "name": "cfn-guard",
          "organization": "Amazon Web Services",
          "semanticVersion": "x.x.x",
          "shortDescription": {
            "text": "AWS CloudFormation Guard is an open-source general-purpose policy-as-code evaluation tool. It provides developers with a simple-to-use, yet powerful and expressive domain-specific language (DSL) to define policies and enables developers to validate JSON- or YAML- formatted structured data with those policies.",
          },
        },
      },
    },
  ],
  "version": "x.x.x",
}
`;
